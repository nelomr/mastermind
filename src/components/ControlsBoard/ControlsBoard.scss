.controls-board {
  display: flex;
  justify-content: space-between;
  align-items: center;
  flex-direction: column;
}

.controls-board--butons {
  display: flex;
  justify-content: center;

  @include breakpoint($min-desktop) {
    flex-direction: column;
  }
}

.controls-board--buton {
  margin-top: 30px;

  @include breakpoint($min-desktop) {
    margin-top: 100px;
  }
}

.controls-board--status {
  margin-top: 30px;
  font-size: 30px;
  font-weight: bold;
  color: $red-color;

  &.is-won {
    color: $green-color;
  }
}

.controls-board--status-title {
  margin-bottom: 30px;
}

.controls-board--status-text {
  font-size: 18px;
  color: $text-color;
}

.controls-board--code {
  display: inline-flex;
  width: 20px;
  height: 20px;
  margin-right: 5px;
  border-radius: 50%;
}