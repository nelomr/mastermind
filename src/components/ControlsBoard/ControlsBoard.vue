<template>
  <div class="controls-board">
    <div 
      v-if="game && status == 'running'"
      class="controls-board--buttons"
    >
      <button 
        @click="sentCheckGuesses()"
        class="controls-board--check secondary-button"
        :disabled="!enableCheck"
      >
        <IconCheck />
      </button>
      <button 
        @click="resetRow()"
        class="controls-board--reset secondary-button"
        :disabled="codeLength == 0"
      >
        <IconRefresh />
      </button>
    </div>

    <button 
      @click.prevent="createGame(), reset()"
      class="controls-board--button primary-button"
    >
      New game
    </button>

    <div 
      v-if="manageStatus"
      :class="{'is-won': status === 'won'}"
      class="controls-board--status">
        <div class="controls-board--status-title" v-text="manageTextStatus"></div>
        <div class="controls-board--status-text" v-text="'Solution'"></div>
        <div
          v-for="(code, index) in game.secret_code"
          :key="'color'+index"
          :style="`background-color: ${ code };`"
          class="controls-board--code"
        />
    </div>
  </div>
</template>

<script>
import EventBus from '@/modules/event-bus';
import { mapState, mapActions } from 'vuex';
import IconCheck from '@/components/icons/IconCheck';
import IconRefresh from '@/components/icons/IconRefresh';

export default {
    name: 'ControlsBoard',
    components: {
      IconCheck,
      IconRefresh
    },
    computed: {
      ...mapState([
        'code',
        'game',
        'status'
      ]),
      manageStatus() {
        return this.status === 'won' || this.status === 'lost'
      },
      manageTextStatus() {
        return this.status === 'won' ? 'You win': (this.status === 'lost') ? 'You lose': '';
      },
      numberSlots() {
        return this.game.num_slots;
      },
      codeLength() {
        return this.code.length;
      },
      enableCheck() {
        let checkUndefineds = this.code.filter(color => typeof color === 'undefined').length;
        return (checkUndefineds == 0 && this.codeLength == 4);
      }
    },
    methods: {
    ...mapActions([
      'createGame',
      'sentCheckGuesses',
      'resetCode'
    ]),
    reset() {
      EventBus.$emit('reset');
    },
    resetRow() {
      EventBus.$emit('resetRow');
      this.resetCode();
    }
  }
}
</script>
<style lang="scss">
  @import "./ControlsBoard.scss";
</style>