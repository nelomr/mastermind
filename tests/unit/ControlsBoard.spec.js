import Vuex from 'vuex';
import { shallowMount, createLocalVue } from '@vue/test-utils';
import ControlsBoard from '@/components/ControlsBoard/ControlsBoard.vue';

const localVue = createLocalVue();
localVue.use(Vuex);

describe('ControlsBoard.vue', () => {
  let game = null;

  let store;
  let actions;
  let computed;

  beforeEach(() => {
    actions = {
      createGame: jest.fn()
    };
    computed = {
      game() {
          return game;
      }
    };
    store = new Vuex.Store({
      actions
    });
  });
  const build = () => {
    const wrapper = shallowMount(ControlsBoard, {
      computed,
      mocks: {$store: store}
    });

    return {
        wrapper,
        buttonNewGame: () => wrapper.find('.controls-board--button')
    };
  };

  it('renders the component no button list', () => {
    const {wrapper} = build();

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('On button click call create game method', () => {
    const {wrapper, buttonNewGame} = build();
    const reset = jest.spyOn(wrapper.vm, 'reset');

    buttonNewGame().trigger('click');
    expect(reset).toHaveBeenCalled();
    expect(actions.createGame).toHaveBeenCalled();
  });

  it('On button click call return game object', () => {
    const {wrapper, buttonNewGame} = build();

    expect(typeof wrapper.vm.game).toEqual(
      expect.objectContaining(null)
    );

    buttonNewGame().trigger('click');
    expect(actions.createGame).toHaveBeenCalled();
    expect(typeof wrapper.vm.game).toBe('object');
  });
})
