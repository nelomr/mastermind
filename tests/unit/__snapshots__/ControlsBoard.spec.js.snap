// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ControlsBoard.vue renders the component no button list 1`] = `
<div class="controls-board">
  <!----> <button class="controls-board--button primary-button">
    New game
  </button>
  <!---->
</div>
`;
